### PR TITLE
[WIP] Use Result for from_storeid

### LINKED ISSUE #778: Rewrite: libimagdiary: FromStoreId -> Option<_> to use Result<_>
https://matthiasbeyer.github.io/imag/src/libimagdiary/diaryid.rs.html#178

We should rewrite this to use `Result<_>` and return proper errors.

Would also be a nice experience for a beginner, though not a complete rust-beginner, I guess.
I guess this would also include some fixes in `imag-diary`, though I'm not completely sure.

### PR DESCRIPTION
**Pull request by rnestler**
This is still work in progress, but should fix #778.

### CODE REVIEW

File: libimagdiary/src/diaryid.rs [hunk 1]
```diff
@@ -212,25 +215,37 @@ impl FromStoreId for DiaryId {
...
         };
 
-        let day   :Option<u32> = cmps.next().and_then(component_to_str).and_then(|s| FromStr::from_str(s).ok());
-        let month :Option<u32> = cmps.next().and_then(component_to_str).and_then(|s| FromStr::from_str(s).ok());
-        let year  :Option<i32> = cmps.next().and_then(component_to_str).and_then(|s| FromStr::from_str(s).ok());
-        let name       = cmps.next().and_then(component_to_str).map(String::from);
+        let day: Result<u32,_> = cmps.next().and_then(component_to_str)
+            .ok_or(DE::new(DEK::ParseError, None))
+            .and_then(|s| FromStr::from_str(s)
+                      .map_err(|e| DE::new(DEK::ParseError, None))); // Some(Box::new(e)))));
```
Comment (rnestler) [inline libimagdiary/src/diaryid.rs]: I have some trouble here boxing the underlying error returned by  `FromStr::from_str(s)`:
```
error[E0277]: the trait bound `std::error::Error: std::marker::Sized` is not satisfied
   --> src/diaryid.rs:228:27
    |
228 |             .and_then(|s| FromStr::from_str(s)
    |                           ^^^^^^^^^^^^^^^^^^^^
    |
    = note: `std::error::Error` does not have a constant size known at compile-time
    = note: required by `std::result::Result`

error[E0277]: the trait bound `std::error::Error: std::marker::Sized` is not satisfied
   --> src/diaryid.rs:230:24
    |
230 |                       .map_err(|e| DE::new(DEK::ParseError, Some(Box::new(e)))));
    |                        ^^^^^^^
    |
    = note: `std::error::Error` does not have a constant size known at compile-time
```

Comment (rnestler) [inline libimagdiary/src/diaryid.rs]: I figured it out. If i use parse instead of FromStr it works. Probably some inference problems.

**Comment by rnestler** [comment]
Just tell me if I need to squash the commits or need to edit the commit messages. I pushed them a bit in a hurry.

File: libimagdiary/src/diaryid.rs [hunk 2]
```diff
@@ -175,62 +178,77 @@ impl Into<NaiveDateTime> for DiaryId {
...
 use std::path::Component;
 
-fn component_to_str<'a>(com: Component<'a>) -> Option<&'a str> {
+fn component_to_str<'a>(com: Component<'a>) -> Result<&'a str, DE> {
     match com {
         Component::Normal(s) => Some(s),
-        _ => None
+        _ => None,
     }.and_then(|s| s.to_str())
+    .ok_or(DE::new(DEK::ParseError, None))
```
Comment (matthiasbeyer) [inline libimagdiary/src/diaryid.rs]: You can use `ok_or(DEK::ParseError.into_error())` for a bit of convenience here.

File: libimagdiary/src/diaryid.rs [hunk 3]
```diff
@@ -175,62 +178,77 @@ impl Into<NaiveDateTime> for DiaryId {
...
-    fn from_storeid(s: &StoreId) -> Option<DiaryId> {
+    fn from_storeid(s: &StoreId) -> Result<DiaryId, DE> {
         use std::str::FromStr;
 
+        use std::path::Components;
+        use std::iter::Rev;
+
+        fn next_component<'a>(components: &'a mut Rev<Components>) -> Result<&'a str, DE> {
+            components.next()
+                .ok_or(DE::new(DEK::ParseError, None))
```
Comment (matthiasbeyer) [inline libimagdiary/src/diaryid.rs]: Also here: `::into_error()`!

File: libimagdiary/src/diaryid.rs [hunk 4]
```diff
@@ -175,62 +178,77 @@ impl Into<NaiveDateTime> for DiaryId {
...
+
+            match (hour, minute) {
+                (Some(h), Some(m)) => Ok((h, m)),
+                _ => return Err(DE::new(DEK::ParseError, None)),
+            }
+        }));
+
+        let day: Result<u32,_> = next_component(&mut cmps)
+            .and_then(|s| s.parse::<u32>()
+                      .map_err(|e| DE::new(DEK::ParseError, Some(Box::new(e)))));
```
Comment (matthiasbeyer) [inline libimagdiary/src/diaryid.rs]: You can use `.map_err_into(DEK::ParseError)`, that will do the rest of the magic automatically. You might need to import `use error::MapErrInto` for this.

File: libimagdiary/src/diaryid.rs [hunk 5]
```diff
@@ -175,62 +178,77 @@ impl Into<NaiveDateTime> for DiaryId {
...
+
+        let month: Result<u32,_> = next_component(&mut cmps)
+            .and_then(|s| s.parse::<u32>()
+                      .map_err(|e| DE::new(DEK::ParseError, Some(Box::new(e)))));
```
Comment (matthiasbeyer) [inline libimagdiary/src/diaryid.rs]: Same here.

File: libimagdiary/src/diaryid.rs [hunk 6]
```diff
@@ -175,62 +178,77 @@ impl Into<NaiveDateTime> for DiaryId {
...
+
+        let year: Result<i32,_> = next_component(&mut cmps)
+            .and_then(|s| s.parse::<i32>()
+                      .map_err(|e| DE::new(DEK::ParseError, Some(Box::new(e)))));
```
Comment (matthiasbeyer) [inline libimagdiary/src/diaryid.rs]: And here.

File: libimagdiary/src/iter.rs [hunk 7]
```diff
@@ -99,9 +99,9 @@ impl<'a> Iterator for DiaryEntryIterator<'a> {
             if next.is_in_diary(self.name) {
                 debug!("Seems to be in diary: {:?}", next);
                 let id = match DiaryId::from_storeid(&next) {
-                    Some(i) => i,
-                    None => {
-                        debug!("Couldn't parse {:?} into DiaryId", next);
+                    Ok(i) => i,
+                    Err(e) => {
+                        debug!("Couldn't parse {:?} into DiaryId: {:?}", next, e);
```
Comment (matthiasbeyer) [inline libimagdiary/src/iter.rs]: Keep the `debug!()` call and also add a `trace_error(&e)` here (with ` uselibimagerror::trace::trace_error;`).

File: libimagdiary/src/diaryid.rs (hunk 4 above)
Comment (rnestler) [inline libimagdiary/src/diaryid.rs]: Done.

File: libimagdiary/src/diaryid.rs (hunk 5 above)
Comment (rnestler) [inline libimagdiary/src/diaryid.rs]: Done.

File: libimagdiary/src/diaryid.rs (hunk 6 above)
Comment (rnestler) [inline libimagdiary/src/diaryid.rs]: Done.

File: libimagdiary/src/iter.rs (hunk 7 above)
Comment (rnestler) [inline libimagdiary/src/iter.rs]: Done. I hope I understood you correctly, I just added the call to trace_error()?

**Comment by rnestler** [comment]
Sorry for not updating this PR for such a long time. I just forgot about it 🙈

File: libimagdiary/src/diaryid.rs (hunk 2 above)
Comment (rnestler) [inline libimagdiary/src/diaryid.rs]: Done.

File: libimagdiary/src/error.rs [hunk 8]
```diff
@@ -27,7 +27,8 @@ generate_error_module!(
         PathConversionError => "Error while converting paths internally",
         EntryNotInDiary     => "Entry not in Diary",
         IOError             => "IO Error",
-        ViewError           => "Error viewing diary entry"
+        ViewError           => "Error viewing diary entry",
+        ParseError          => "Error while parsing"
```
Comment (matthiasbeyer) [inline libimagdiary/src/error.rs]: Maybe name it `IdParseError => "Error while parsing ID"`.

Besides that it really looks awesome and I'm looking forward to push the merge button :smile:

**Comment by matthiasbeyer** [comment]
Thanks a lot for your PR! :smile: :+1:

**Comment by rnestler** [comment]
> Thanks a lot for your PR! 😄 👍

No problem, was fun hacking on it 😃

**Pull request merged**